### PR TITLE
[Numpy] Add infra for performing constraint check

### PR DESCRIPTION
**Pull request by xidulu**
## Description ##
https://github.com/apache/incubator-mxnet/issues/16868
To perform constraint check on a tensor, one only need to multiply `npx.constraint_check(condition(tensor), msg)` with the origin tensor. In such way, the `constraint_check` is forced to be evaluated.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at https://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by leezu** [comment]
Could you improve the documentation to clarify how to ensure that the constraint check is executed?

**Comment by xidulu** [comment]
> Could you improve the documentation to clarify how to ensure that the constraint check is executed?

@leezu 
I add some brief explanation and a toy example to demonstrate how this OP work in the doc section.
Unfortunately, I cannot provide further technical details regarding why this `hack` could work, cuz I don't have know much about the computation graph engine.
But according to my test cases, this implementation could now do what it is designed to do.
https://github.com/apache/incubator-mxnet/pull/17272/files#diff-23804c433abe5a6ae5c3a113425d3df8R60

File: src/operator/numpy/np_constraint_check.cc [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+namespace mxnet {
+namespace op {
+
+template<>
+void GetReduceOutput<cpu>(mshadow::Stream<cpu> *s, const TBlob &output_blob, bool *red_output) {
+  *red_output = static_cast<bool>(*output_blob.dptr<bool>());
+}
+
+inline bool ConstraintCheckShape(const nnvm::NodeAttrs& attrs,
+                        std::vector<TShape>* in_attrs,
```
Comment (haojin2) [inline src/operator/numpy/np_constraint_check.cc]: alignment.

File: src/operator/numpy/np_constraint_check.cc [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+  CHECK_EQ(out_attrs->size(), 1U);
+  if (!shape_is_known(in_attrs->at(0))) {
+    return false;
+  }
+  SHAPE_ASSIGN_CHECK(*out_attrs, 0, TShape(0, -1))
+  return true;
+}
+
+inline bool ConstraintCheckType(const nnvm::NodeAttrs& attrs,
+                       std::vector<int>* in_attrs,
```
Comment (haojin2) [inline src/operator/numpy/np_constraint_check.cc]: alignment

File: src/operator/numpy/np_constraint_check.cc [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+  CHECK_EQ(in_attrs->size(), 1U);
+  CHECK_EQ(out_attrs->size(), 1U);
+  CHECK(in_attrs->at(0) == mshadow::kBool);
+  TYPE_ASSIGN_CHECK(*out_attrs, 0, mshadow::kBool);
+  return out_attrs->at(0) != -1 && in_attrs->at(0) != -1;
+}
+
+DMLC_REGISTER_PARAMETER(ConstraintCheckParam);
+
+NNVM_REGISTER_OP(_npx_constraint_check)
```
Comment (haojin2) [inline src/operator/numpy/np_constraint_check.cc]: If you do not want this op to be exposed to users directly, `_npi_` is recommended.

File: src/operator/numpy/np_constraint_check.h [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  DMLC_DECLARE_PARAMETER(ConstraintCheckParam) {
+    DMLC_DECLARE_FIELD(msg)
+    .set_default("Constraint violated.")
+    .describe("Error message raised when constraint violated");
+  }
+};
+
+template <typename xpu>
+void ConstraintCheckForward(const nnvm::NodeAttrs& attrs, const OpContext& ctx,
+                  const std::vector<TBlob>& inputs,
```
Comment (haojin2) [inline src/operator/numpy/np_constraint_check.h]: alignment.

**Pull request merged**